File: systems-engineering/operating-systems/impl/io_model_lab/java/EchoServer.java
import java.io.IOException;
import java.net.InetSocketAddress;
import java.nio.ByteBuffer;
import java.nio.channels.SelectionKey;
import java.nio.channels.Selector;
import java.nio.channels.ServerSocketChannel;
import java.nio.channels.SocketChannel;
import java.util.Iterator;
import java.util.Set;

/**
 * Non-blocking echo server using Java NIO Selector.
 * On Linux, Selector uses epoll under the hood (sun.nio.ch.EPollSelectorImpl).
 */
public class EchoServer {
    public static void main(String[] args) throws IOException {
        Selector selector = Selector.open();
        ServerSocketChannel serverSocket = ServerSocketChannel.open();
        serverSocket.bind(new InetSocketAddress("0.0.0.0", 8080));
        serverSocket.configureBlocking(false);
        serverSocket.register(selector, SelectionKey.OP_ACCEPT);

        System.out.println("NIO echo server on :8080");

        while (true) {
            selector.select();
            Set<SelectionKey> selectedKeys = selector.selectedKeys();
            Iterator<SelectionKey> iter = selectedKeys.iterator();
            while (iter.hasNext()) {
                SelectionKey key = iter.next();
                iter.remove();

                if (key.isAcceptable()) {
                    SocketChannel client = serverSocket.accept();
                    client.configureBlocking(false);
                    client.register(selector, SelectionKey.OP_READ);
                    System.out.println("Accepted " + client.getRemoteAddress());
                }

                if (key.isReadable()) {
                    SocketChannel client = (SocketChannel) key.channel();
                    ByteBuffer buffer = ByteBuffer.allocate(1024);
                    int bytesRead = client.read(buffer);
                    if (bytesRead > 0) {
                        buffer.flip();
                        client.write(buffer);
                    } else if (bytesRead == -1) {
                        System.out.println("Closing " + client.getRemoteAddress());
                        client.close();
                    }
                }
            }
        }
    }
}

File: systems-engineering/operating-systems/impl/io_model_lab/ts/echo_server.ts
import net from "net";

/**
 * Node.js net echo server.
 * Under the hood, libuv uses epoll (Linux) / kqueue (macOS) / IOCP (Windows)
 * to multiplex I/O in a single event loop thread.
 */
const server = net.createServer((socket) => {
  console.log("Accepted", socket.remoteAddress, socket.remotePort);
  socket.on("data", (data) => {
    socket.write(data); // echo
  });
  socket.on("end", () => {
    console.log("Closing", socket.remoteAddress, socket.remotePort);
  });
});

server.listen(8080, () => {
  console.log("libuv echo server on :8080");
  console.log("libuv backend:", process.platform === "linux" ? "epoll" : process.platform === "darwin" ? "kqueue" : "iochcp");
});
